jib-core/src/main/java/com/google/cloud/tools/jib/event/progress/AllocationCompletionTracker.java

```diff
@@ -112,56 +115,67 @@ boolean updateProgress(Allocation allocation, long units) {
-    while (!unfinishedInsertionOrderUnits.isEmpty()) {
-      unfinishedAllocations.add(unfinishedInsertionOrderUnits.remove().allocation);
-    }
-    return unfinishedAllocations.build();
+  @VisibleForTesting
+  List<Allocation> getUnfinishedAllocations() {
+    return completionMap
+        .values()
+        .stream()
+        .filter(IndexedRemainingUnits::isUnfinished)
+        .sorted()
+        .map(remainingUnits -> remainingUnits.allocation)
+        .collect(Collectors.toList());
   }
 
   /**
-   * Helper method for {@link #updateProgress(Allocation, long)}. Adds units to {@code
-   * insertionOrderUnits}. Updates {@link InsertionOrderUnits} for parent {@link Allocation}s if
-   * {@code insertionOrderUnits.units} reached completion (equals {@code
-   * allocation.getAllocationUnits()}. This method is <em>not</em> thread-safe for the {@code
-   * insertionOrderUnits} and should be called atomically relative to the {@code
-   * insertionOrderUnits}.
+   * Helper method for {@link #updateProgress(Allocation, long)}. Subtract {@code units} from {@code
+   * indexedRemainingUnits}. Updates {@link IndexedRemainingUnits} for parent {@link Allocation}s if
+   * remaining units becomes 0. This method is <em>not</em> thread-safe for the {@code
+   * indexedRemainingUnits} and should be called atomically relative to the {@code
+   * indexedRemainingUnits}.
    *
-   * @param insertionOrderUnits the {@link InsertionOrderUnits} to update progress for
+   * @param indexedRemainingUnits the {@link IndexedRemainingUnits} to update progress for
    * @param units the units of progress
    */
-  private void updateInsertionOrderUnits(InsertionOrderUnits insertionOrderUnits, long units) {
-    Allocation allocation = insertionOrderUnits.allocation;
+  private void updateIndexedRemainingUnits(
+      IndexedRemainingUnits indexedRemainingUnits, long units) {
+    Allocation allocation = indexedRemainingUnits.allocation;
 
-    long newUnits = insertionOrderUnits.units.addAndGet(units);
-    if (newUnits > allocation.getAllocationUnits()) {
+    long newUnits = indexedRemainingUnits.remainingUnits.addAndGet(-units);
+    if (newUnits < 0L) {
       throw new IllegalStateException(
           "Progress exceeds max for '"
               + allocation.getDescription()
               + "': "
-              + newUnits
-              + " > "
+              + -newUnits
+              + " more beyond "
               + allocation.getAllocationUnits());
     }
 
     // Updates the parent allocations if this allocation completed.
-    if (newUnits == allocation.getAllocationUnits()) {
+    if (newUnits == 0L) {
       allocation
           .getParent()
           .ifPresent(
               parentAllocation ->
-                  updateInsertionOrderUnits(
+                  updateIndexedRemainingUnits(
                       Preconditions.checkNotNull(completionMap.get(parentAllocation)), 1L));
     }
   }
+
+  ImmutableList<String> getUnfinishedLeafTasks() {
```

Comment:
Note to reviewers: this is moved from `ProgressDisplayGenerator.getLeafAllocations()`, fixing the TODO in that method. (#1356). Check the original method first.